### PR TITLE
feat: optimizations and bug fixes

### PR DESCRIPTION
**Pull request by kobigurk**
* Optimizes gas usage by removing the signals tree, replacing with a mapping.
* Fixes nullifier bug by upgrading to latest snarkjs. Adds negative tests for the bug.
* Saves all the leaves.
* Upgrades to Solidity 0.5.10.

### CODE REVIEW

**Comment by kobigurk** [comment]
@weijiekoh can you please review?

File: semaphorejs/contracts/Ownable.sol [hunk 1]
```diff
@@ -1,4 +1,4 @@
-pragma solidity ^0.4.24;
+pragma solidity ^0.5.0;
```
Comment (weijiekoh) [inline semaphorejs/contracts/Ownable.sol:1]: I suggest upgrading to the OZ 2.0 Ownable contract.

I can do this in a separate PR as I'd like to add revert messages too.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v2.3.0/contracts/ownership/Ownable.sol

**Comment by weijiekoh** [comment]
`MerkleTreeTester.sol` needs upgrading to 0.5.x too

**Comment by weijiekoh** [comment]
The cached `verifier.sol` is causing the CI build to fail.

**Comment by kobigurk** [comment]
> The cached `verifier.sol` is causing the CI build to fail.

Ah! That's what it is. Thanks.

**Comment by kobigurk** [comment]
Fixes https://github.com/kobigurk/semaphore/issues/16.

File: semaphorejs/contracts/Ownable.sol (hunk 1 above)
Comment (kobigurk) [reply semaphorejs/contracts/Ownable.sol:1]: Sounds good, let's leave it to a separate PR.

**Pull request merged**